### ISSUE TITLE
vim.diagnostic.disable doesn't fully disable the diagnostic

### ISSUE THREAD

**Issue by sidequestboy** [neovim/neovim#25131]
### Problem

While the virtualtext does disappear, it does not affect the output of `get_namespaces`, `get`, `get_next`, `get_next_pos`, `get_prev`, `get_prev_pos`, `goto_next`, `goto_prev`, `setloclist`, `setqflist`.

Which means we have to resort to some gymnastics to manually filter per-namespace or per-diagnostic with `is_disabled`.

### Steps to reproduce

In lspconfig config, I disable tsserver diagnostics for javascript files:

```lua
local on_attach = function(client, bufnr)
  if client.name == 'tsserver' and (vim.bo[bufnr].filetype == 'javascript' or vim.bo[bufnr].filetype == 'javascriptreact') then
    local ns = vim.lsp.diagnostic.get_namespace(client.id)
    vim.diagnostic.disable(bufnr, ns)
  end
  local nmap = function(keys, func, desc)
    if desc then
      desc = 'LSP: ' .. desc
    end

    vim.keymap.set('n', keys, func, { buffer = bufnr, desc = desc })
  end
end
local servers = {
  tsserver = {},
  eslint = {},
}
-- nvim-cmp supports additional completion capabilities, so broadcast that to servers
local capabilities = vim.lsp.protocol.make_client_capabilities()
capabilities = require('cmp_nvim_lsp').default_capabilities(capabilities)

-- Ensure the servers above are installed
local mason_lspconfig = require 'mason-lspconfig'

mason_lspconfig.setup {
  ensure_installed = vim.tbl_keys(servers),
}

mason_lspconfig.setup_handlers {
  function(server_name)
    require('lspconfig')[server_name].setup {
      capabilities = capabilities,
      on_attach = on_attach,
      settings = servers[server_name],
      filetypes = (servers[server_name] or {}).filetypes,
    }
  end,
}
```

But I have to jump through hoops to get all the functions listed above to not show me disabled diagnostics.

### Expected behavior

The listed functions should adjust their output according to disabled namespaces, or at least have an option to only include enabled diagnostics.

### Neovim version (nvim -v)

0.9.2

### Vim (not Nvim) behaves the same?

no

### Operating system/version

macOS 14.0

### Terminal name/version

kitty

### $TERM environment variable

tmux-256color

### Installation

homebrew

**Comment by MariaSolOs** [comment]
I might be wrong here but this seems to be by design? `vim.diagnostic.disable` just calls `vim.diagnostic.hide`, which "only clears the decorations displayed in the buffer" and makes sure that further calls to `vim.diagnostic.show` do not re-enable the decorations again.

@sidequestboy it sounds like you want to use `vim.diagnostic.reset`, which actually removes diagnostics and not only the decorations.

**Comment by gpanders** [comment]
`vim.diagnostic.disable()` and `vim.diagnostic.enable()` only affect the diagnostic UI elements (virtual text, signs, etc.). It does not affect the underlying diagnostic data at all.

`vim.diagnostic.get()` already accepts a few options for filtering diagnostics. Adding an `enabled` key to filter out disabled diagnostics doesn't seem unreasonable, though a better approach would probably be to make all of these functions accept a generic `filter` function predicate (that way we don't need special `severity`, `enabled`, `namespace`, etc. options).

**Comment by sidequestboy** [comment]
thanks for the suggestion Maria, `reset` seems to get rid of the diagnostics from `goto_next`, quickfix list etc, however it's temporary: as soon as another edit is made, the diagnostics from that namespace show up again. Another side effect of `reset` is that code actions from that namespace disappear also.

I suppose this could be used in combination with `disable` to hide and also affect the output of those other commands (e.g. call `reset` before calling `goto_next`), but even this isn't what I want in this case, since I want tsserver's code actions without its diagnostics for javascript buffers

**Comment by sidequestboy** [comment]
yes something like that would be great gpanders, although personally I would expect a "disabled" diagnostic to just not show up in the quickfix list, goto_next, get, etc. by default.

**Comment by MariaSolOs** [comment]
@sidequestboy note that some code actions are tied to diagnostics (they're suggestions on how to fix the errors), so when disabling diagnostics I'm not surprised that code actions also disappear. You might still get TSServer's refactors (which are defined [here](https://github.com/microsoft/TypeScript/tree/9cbcf010ce0701a25f01a2a074000db34f80cc17/src/services/refactors)), but just a heads up.

**Comment by sidequestboy** [comment]
That's fair, also to be clear, they don't disappear when `disable` is called, but they do disappear temporarily when `reset` is called

**Comment by gh-liu** [comment]
`vim.diagnostic.jump`, `vim.diagnostic.setqflist`, `vim.diagnostic.open_float`, `vim.diagnostic.status` seems still take disabled diagnostic into account.

https://github.com/neovim/neovim/blob/088ba835ed5f89869a30bc13e8d203894328a69d/runtime/lua/vim/diagnostic.lua#L889-L892

if the `opt.enabled` not be set, it will treat all diagnostics as enabled，which is a bit of a counter-intuitive, cause I already disable the diagnostic.

the doc says: if enabled is nil, enablement is ignored. I still confused why ignored it? I disabled the diagnostics, I can not see that, but I could jump to it, I could open the diagnostic float window, and the `status()` count the diagnostic.
```
      • {enabled}?    (`boolean`, default: `nil`) Limit diagnostics to only
                      enabled or disabled. If nil, enablement is ignored. See
                      |vim.diagnostic.enable()|
``

**Comment by MariaSolOs** [comment]
@gh-liu https://github.com/neovim/neovim/pull/33981 was meant to preserve the existing behavior. If you want to filter out disabled diagnostics set `enabled = true`.
